compiler/src/iree/compiler/Codegen/Common/VectorReductionToGPU.cpp

```diff
@@ -105,6 +156,78 @@ static Attribute getCombiningKindIdentity(OpBuilder &builder,
   return Attribute();
 }
 
+/// Compute the value on a single thread to get per lane reduction value.
+/// If bit-width is not supported on shuffle operations, and a lower precision,
+/// we represent them as a vector S.T we can pack them into a single 32-bit
+/// width for shuffles.
+static Value reduceToSupportedWidth(Location loc, OpBuilder &builder,
+                                    Value input, vector::CombiningKind kind) {
+  auto vecType = input.getType().cast<VectorType>();
+  Type elementType = vecType.getElementType();
+  int64_t vecSize = vecType.getShape()[0];
+  unsigned bitWidth = elementType.getIntOrFloatBitWidth();
+  // Simply reduce if it's already 32 bits.
+  if (bitWidth == kShuffleBitWidth) {
+    return builder.create<vector::ReductionOp>(loc, kind, input);
+  }
+  assert(kShuffleBitWidth % bitWidth == 0 &&
+         "Bitwidth needs to be able to be packed into shuffle-bitwidth.");
+  int64_t unrollCount = kShuffleBitWidth / bitWidth;
+  // Original size needs to be divisble by or less than unroll count to
+  // determine slice size.
+  assert(vecSize % unrollCount == 0 || vecSize < unrollCount);
+  unsigned sliceSize = vecSize / unrollCount;
+  VectorType unrolledLaneValType = VectorType::get({unrollCount}, elementType);
+  Value perLaneReduction = builder.create<arith::ConstantOp>(
+      loc, builder.getZeroAttr(unrolledLaneValType));
+  if (vecSize % unrollCount == 0) {
+    // Unroll reductions s.t we can pack into a supported 32-bitWidth format.
+    for (int64_t i = 0; i < unrollCount; i++) {
+      Value laneValSlice = builder.create<vector::ExtractStridedSliceOp>(
+          loc, input,
+          /*offsets=*/ArrayRef<int64_t>{sliceSize * i},
+          /*sizes=*/ArrayRef<int64_t>{sliceSize},
+          /*strides=*/ArrayRef<int64_t>{1});
+      Value reductionSlice =
+          builder.create<vector::ReductionOp>(loc, kind, laneValSlice);
+      SmallVector<int64_t> perLaneUnrollId = {i};
+      perLaneReduction = builder.create<vector::InsertOp>(
+          loc, reductionSlice, perLaneReduction, perLaneUnrollId);
+    }
+  } else {
+    // In cases where vecSize < unrollCount, we would pad the vector
+    // with identity elements until it's total bit size is 32.
+    Attribute identityAttr =
```

Comment:
is there a test for this case? Can we add one?

Reply:
digging up for the dispatches by looking through the dispatches of SD 😅

Reply:
done, simple_half_reduce_small is supposed to represent it :)